### PR TITLE
Graph parsing

### PR DESCRIPTION
**Pull request by wilmhit**

### CODE REVIEW

File: graph-coloring/graph_parsing.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    """
+
+    graph = []
+    lines = split_lines(graph_file)
+    for line in lines:
+        cells = split_spaces(line)
+        cells = remove_empty_cells(cells)
+        graph.append(parse_into_numbers(cells))
+    return graph
+
```
Comment (songofroland) [inline graph-coloring/graph_parsing.py:16]:
```
    graph = []
    for line in graph_file.split("\n"):
        cells = line.split(" ") 
        # I don't get what's remove_empty_cells for
        graph.append([int(cell) for cell in cells])
    return graph
```
We don't need those functions. Python is idiomatic on its own.

To consider: using numpy (the question is if we'll take advantage of it - 16MB)
```
lines = graph_file.split("\n")
graph = [list(numpy.fromstring(line, dtype=int, sep="  ")) for line in lines]
```

File: graph-coloring/types.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+
+Graph = list[list[int]]
```
Comment (songofroland) [inline graph-coloring/types.py:2]: a big + for setting up static analysis tool

**Review by songofroland — CHANGES_REQUESTED** [review]

File: Pipfile [hunk 3]
```diff
@@ -0,0 +1,13 @@
+[[source]]
+url = "https://pypi.org/simple"
+verify_ssl = true
+name = "pypi"
+
+[packages]
+
+[dev-packages]
+yapf = "*"
```
[the remark this answers is not in the document]
Comment (wilmhit) [reply Pipfile:9]: pep8. It's default

File: graph-coloring/__main__.py [hunk 4]
```diff
@@ -1,14 +1,18 @@
...
+def main(args: Namespace) -> None:
     graph_file = get_graph_file(args.GRAPH)
-    output = process_graph(graph_file)
+    graph = parse_graph_file(graph_file)
+    output = process_graph(graph)
     save_png(output, args.out)
 
 
-def process_graph(graph) -> None:
+def process_graph(graph: Graph) -> Idk:
```
[the remark this answers is not in the document]
Comment (wilmhit) [reply graph-coloring/__main__.py:15]: It will be changed once this does something.

File: graph-coloring/graph_parsing.py (hunk 1 above)
Comment (wilmhit) [reply graph-coloring/graph_parsing.py:16]: Numpy is definitely overkill for this project. Empty cells appear when there is double space in  input file.

File: graph-coloring/file_operations.py [hunk 5]
```diff
@@ -1,7 +1,8 @@
+from customtypes import Idk
```
Comment (songofroland) [inline graph-coloring/file_operations.py:1]: typo in import I guess

File: graph-coloring/graph_parsing.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+from customtypes import Graph
+
+
+def graph_from_string(graph_file: str) -> Graph:
+    """String must be separated with spaces and \n"""
```
Comment (songofroland) [inline graph-coloring/graph_parsing.py:5]:
```
The string should represent an adjacency matrix, separated with spaces and "\n",
e.g.
0 1 1 \n
1 0 1 \n
1 1 0 \n
```

**Review by songofroland — APPROVED** [review]

**Pull request merged**